### DISCUSSION TITLE
Passing arguments to DAP

### DISCUSSION THREAD

**Discussion by ranjithshegde** [mfussenegger/nvim-dap#420] [Q&A] [answered]
I am trying to debug my cpp program that requires to be run on dGPU (for OpenCL). When I debug with gdb i would just call gdb with my dGPU render-offload prefix `prime-run gdb myProgramBinary`. When I execute `run` in gdb it runs as expected on dGPU

Is there a way to do this with nvim-dap? I am using vscode-cpptools with Gdb-MI.

What I have tried so far is pre-pending prime-run in cpp-tools adapter settings
```lua
dap.adapters.prime = {
-- stuff
command = "/usr/bin/prime-run " .. actualDAPCommand
```
which fails to launch. I have also tried prepending the `prime-run` in the configurations through `program = function() "/usr/bin/prime-run " .. whateverBinaryFetchingMethod` also with no luck.

Is there a way to declare this?

**Comment by mfussenegger** [comment, marked as answer]
Did you try using the `miDebuggerPath` and `miDebuggerArgs` options from [vscode-cpptools launch.json reference](https://code.visualstudio.com/docs/cpp/launch-json-reference#_customizing-gdb-or-lldb)?

The command within the adapter configuration needs to start the debug-adapter, not the debugger. As shown in the dap.txt help:

```
DAP-Client ----- Debug Adapter ------- Debugger ------ Debugee
(nvim-dap)  |   (per language)  |   (per language)    (your app)
            |                   |
            |        Implementation specific communication
            |        Debug-adapter and debugger could be the same process
            |
     Communication via debug adapter protocol

```

The debug adapter in your case is `vscode-cpptools` and the debugger is `gdb`.

`dap.adapters` defines how to start the debug adapter
`dap.configurations` defines how to start the debugee. Depending on the debug adapter, it can also influence how the debugger is started. Looks like this is possible with vscode-cpptools.

**Reply by ranjithshegde** [reply]
Thank you for the explaination of the process. I understand now. Playing around with `miDebuggerPath` did work for me.

Just for others looking to perpend/append commands to the debugger, I did it with a shell script 
```sh
#!/bin/sh
MY_FLAGS myPreviousCommands path/to/debugger "$@"
```
then setting path to that script as `miDebuggerPath`
